python/dgl/distributed/dist_dataloader.py

```diff
@@ -0,0 +1,111 @@
+import multiprocessing as mp
+import dgl
+from . import shutdown_servers, finalize_client
+
+DGL_QUEUE_TIMEOUT = 10
+
+__all__ = ["DistDataLoader"]
+
+
+def close():
+    shutdown_servers()
+    finalize_client()
+
+
+def init_fn(collate_fn, mp_queue):
+    global gfn
+    global queue
+    queue = mp_queue
+    gfn = collate_fn
+    import atexit
+    atexit.register(close)
+
+
+def deregister_torch_ipc():
+    from multiprocessing.reduction import ForkingPickler
+    import torch
+    ForkingPickler._extra_reducers.pop(torch.cuda.Event)
+    for t in torch._storage_classes:
+        ForkingPickler._extra_reducers.pop(t)
+    for t in torch._tensor_classes:
+        ForkingPickler._extra_reducers.pop(t)
+    ForkingPickler._extra_reducers.pop(torch.Tensor)
+    ForkingPickler._extra_reducers.pop(torch.nn.parameter.Parameter)
+
+
+def call_collate_fn(next_data):
+    result = gfn(next_data)
+    queue.put(result)
+    return 1
+
+
+class DistDataLoader:
+    """"""
+
+    def __init__(self, dataset, batch_size, collate_fn, num_workers, drop_last, queue_size=None):
+        """"""
+        assert num_workers > 0
+        if queue_size is None:
+            queue_size = num_workers * 4
+        self.queue_size = queue_size
+        self.dataset = dataset
+        self.batch_size = batch_size
+        self.queue_size = queue_size
+        self.collate_fn = collate_fn
+        self.current_pos = 0
+        self.num_workers = num_workers
+        self.m = mp.Manager()
+        self.queue = self.m.Queue(maxsize=queue_size)
+        ctx = mp.get_context("spawn")
+        self.drop_last = drop_last
+        self.expected_idxs = len(dataset) // batch_size
+        if not drop_last and len(dataset) % batch_size != 0:
+            self.expected_idxs += 1
+        self.send_idxs = 0
+        self.recv_idxs = 0
+        self.pool = ctx.Pool(
+            num_workers, initializer=init_fn, initargs=(collate_fn, self.queue))
+        for _ in range(queue_size):
+            self._request_next_batch()
+
+    def __next__(self):
+        self._request_next_batch()
+        if self.recv_idxs < self.expected_idxs:
+            result = self.queue.get(timeout=DGL_QUEUE_TIMEOUT)
```

Comment:
what does this timeout mean? if it time out, what is result?

Reply:
It will raise an error if doesn't get any data from queue until timeout. Usually this would not happen. Sometime maybe server/client failed which made the returned results less than expected, which could trigger this exception

Reply:
do we need the timeout?